File: reference-implementations/cid/cid.py
"""
koad.generate.handle + koad.generate.cid — Python reference implementation
Spec: VESTA-SPEC-027 (CID Privacy Primitive)

Requires: Python 3.6+ (hashlib is stdlib)

Test vectors:
    handle("https://github.com/koad")  → "httpsgithubcomkoad"
    cid("httpsgithubcomkoad")          → "GdYZWjcjY6Y2XonnM"
"""

import hashlib
import re

CHARSET = "23456789ABCDEFGHJKLMNPQRSTWXYZabcdefghijkmnopqrstuvwxyz"
CHARSET_LEN = 55
CID_LEN = 17


def handle(s: str) -> str:
    """Normalize: lowercase, strip non-alphanumeric characters."""
    return re.sub(r"[^a-z0-9]", "", s.lower())


def cid(s: str) -> str:
    """Compute a 17-character opaque CID from any string input."""
    h = handle(s)
    digest = hashlib.sha256(h.encode("utf-8")).digest()
    return "".join(CHARSET[digest[i] % CHARSET_LEN] for i in range(CID_LEN))


if __name__ == "__main__":
    import sys

    if len(sys.argv) < 2 or sys.argv[1] == "test":
        print("=== Test Vectors ===")
        cases = [
            ("handle", "https://github.com/koad", "httpsgithubcomkoad"),
            ("cid",    "httpsgithubcomkoad",        "GdYZWjcjY6Y2XonnM"),
            ("cid",    "koad",                       "TysPFWq8Nr5LZQQnM"),
        ]
        for fn, inp, expected in cases:
            result = handle(inp) if fn == "handle" else cid(inp)
            status = "PASS" if result == expected else "FAIL"
            print(f"{status} {fn}({inp!r}) = {result!r}" +
                  (f"  (want {expected!r})" if result != expected else ""))
    elif sys.argv[1] == "handle":
        print(handle(sys.argv[2]))
    elif sys.argv[1] == "cid":
        print(cid(sys.argv[2]))
    else:
        print(f"Usage: {sys.argv[0]} handle|cid|test <string>", file=sys.stderr)
        sys.exit(1)

File: reference-implementations/cid/cid.go
// koad.generate.handle + koad.generate.cid — Go reference implementation
// Spec: VESTA-SPEC-027 (CID Privacy Primitive)
//
// go run cid.go test
// go run cid.go handle "https://github.com/koad"
// go run cid.go cid "httpsgithubcomkoad"
//
// Test vectors:
//   handle("https://github.com/koad")  → "httpsgithubcomkoad"
//   cid("httpsgithubcomkoad")          → "GdYZWjcjY6Y2XonnM"

package main

import (
	"crypto/sha256"
	"fmt"
	"os"
	"strings"
	"unicode"
)

const (
	charset    = "23456789ABCDEFGHJKLMNPQRSTWXYZabcdefghijkmnopqrstuvwxyz"
	charsetLen = 55
	cidLen     = 17
)

// Handle normalizes a string: lowercase, strip non-alphanumeric characters.
func Handle(s string) string {
	s = strings.ToLower(s)
	var b strings.Builder
	for _, r := range s {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// CID computes a 17-character opaque identifier from any string input.
func CID(s string) string {
	h := Handle(s)
	digest := sha256.Sum256([]byte(h))
	result := make([]byte, cidLen)
	for i := 0; i < cidLen; i++ {
		result[i] = charset[int(digest[i])%charsetLen]
	}
	return string(result)
}

func main() {
	if len(os.Args) < 2 || os.Args[1] == "test" {
		fmt.Println("=== Test Vectors ===")
		type tc struct{ fn, input, want string }
		cases := []tc{
			{"handle", "https://github.com/koad", "httpsgithubcomkoad"},
			{"cid", "httpsgithubcomkoad", "GdYZWjcjY6Y2XonnM"},
			{"cid", "koad", "TysPFWq8Nr5LZQQnM"},
		}
		for _, c := range cases {
			var got string
			if c.fn == "handle" {
				got = Handle(c.input)
			} else {
				got = CID(c.input)
			}
			status := "PASS"
			if got != c.want {
				status = "FAIL"
			}
			fmt.Printf("%s %s(%q) = %q", status, c.fn, c.input, got)
			if got != c.want {
				fmt.Printf("  (want %q)", c.want)
			}
			fmt.Println()
		}
		return
	}

	if len(os.Args) < 3 {
		fmt.Fprintf(os.Stderr, "Usage: %s handle|cid|test <string>\n", os.Args[0])
		os.Exit(1)
	}

	switch os.Args[1] {
	case "handle":
		fmt.Println(Handle(os.Args[2]))
	case "cid":
		fmt.Println(CID(os.Args[2]))
	default:
		fmt.Fprintf(os.Stderr, "Usage: %s handle|cid|test <string>\n", os.Args[0])
		os.Exit(1)
	}
}
